pallets/liquidity-pools-gateway/queue/src/lib.rs

```diff
@@ -0,0 +1,261 @@
+	#[pallet::storage]
+	#[pallet::getter(fn message_queue)]
+	pub type MessageQueue<T: Config> = StorageMap<_, Blake2_128Concat, T::MessageNonce, T::Message>;
+
+	/// Storage for messages that failed during processing.
+	#[pallet::storage]
+	#[pallet::getter(fn failed_message_queue)]
+	pub type FailedMessageQueue<T: Config> =
+		StorageMap<_, Blake2_128Concat, T::MessageNonce, (T::Message, DispatchError)>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub (super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// A message was submitted.
+		MessageSubmitted {
+			nonce: T::MessageNonce,
+			message: T::Message,
+		},
+
+		/// Message execution success.
+		MessageExecutionSuccess {
+			nonce: T::MessageNonce,
+			message: T::Message,
+		},
+
+		/// Message execution failure.
+		MessageExecutionFailure {
+			nonce: T::MessageNonce,
+			message: T::Message,
+			error: DispatchError,
+		},
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Message not found.
+		MessageNotFound,
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_idle(_now: BlockNumberFor<T>, max_weight: Weight) -> Weight {
+			Self::service_message_queue(max_weight)
+		}
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Convenience method for manually processing a message.
+		///
+		/// If the execution fails, the message gets moved to the
+		/// `FailedMessageQueue` storage.
+		#[pallet::weight(T::WeightInfo::process_message())]
+		#[pallet::call_index(0)]
+		pub fn process_message(origin: OriginFor<T>, nonce: T::MessageNonce) -> DispatchResult {
+			ensure_signed(origin)?;
+
+			let message = MessageQueue::<T>::take(nonce).ok_or(Error::<T>::MessageNotFound)?;
+
+			match T::MessageProcessor::process(message.clone()) {
+				Ok(_) => {
+					Self::deposit_event(Event::<T>::MessageExecutionSuccess { nonce, message });
+
+					Ok(())
+				}
+				Err(e) => {
+					FailedMessageQueue::<T>::insert(nonce, (message.clone(), e.error));
+
+					Self::deposit_event(Event::<T>::MessageExecutionFailure {
+						nonce,
+						message,
+						error: e.error,
+					});
+
+					Ok(())
+				}
+			}
+		}
+
+		/// Convenience method for manually processing a failed message.
+		#[pallet::weight(T::WeightInfo::process_failed_message())]
+		#[pallet::call_index(1)]
+		pub fn process_failed_message(
+			origin: OriginFor<T>,
+			nonce: T::MessageNonce,
+		) -> DispatchResult {
+			ensure_signed(origin)?;
+
+			let (message, _) =
+				FailedMessageQueue::<T>::get(nonce).ok_or(Error::<T>::MessageNotFound)?;
+
+			match T::MessageProcessor::process(message.clone()) {
+				Ok(_) => {
+					Self::deposit_event(Event::<T>::MessageExecutionSuccess { nonce, message });
+
+					FailedMessageQueue::<T>::remove(nonce);
+
+					Ok(())
+				}
+				Err(e) => {
+					Self::deposit_event(Event::<T>::MessageExecutionFailure {
+						nonce,
+						message,
+						error: e.error,
+					});
+
+					Ok(())
```

Comment:
Why not make failing the extrinsic here?

Reply:
Returning error on these extrinsics will revert any storage changes that are done during processing and we want to avoid that.